HealthyLLM/HealthyLLM/Fetcher/HealthDataFetcher.swift

```diff
@@ -8,99 +8,191 @@
+            sortedBy: [SortDescriptor(\.startDate, order: .reverse)]
+        ).first
+        let weightSample = try? await healthKit.query(
+            .bodyMass,
+            timeRange: .ever,
+            limit: 1,
+            sortedBy: [SortDescriptor(\.startDate, order: .reverse)]
+        ).first
+        let bmiSample = try? await healthKit.query(
+            .bodyMassIndex,
+            timeRange: .ever,
+            limit: 1,
+            sortedBy: [SortDescriptor(\.startDate, order: .reverse)]
+        ).first
+
+        let sex = try? healthKit.healthStore.biologicalSex().biologicalSex.description
+        let dateOfBirth = try? healthKit.healthStore.dateOfBirthComponents().date
+        let height = heightSample?.quantity.doubleValue(for: .meterUnit(with: .centi)) ?? 0
+        let weight = weightSample?.quantity.doubleValue(for: .gramUnit(with: .kilo)) ?? 0
+        let bmi = bmiSample?.quantity.doubleValue(for: .count()) ?? 0
+
         return .init(
             name: nil,
-            dateOfBirth: try? healthStore.dateOfBirthComponents().date,
-            sex: try? healthStore.biologicalSex().biologicalSex.description,
-            height: height != nil ? "\((height! / 100).rounded())cm" : nil,
-            weight: weight != nil ? "\(weight!.rounded())kg" : nil,
-            bmi: bmi != nil ? "\(bmi!.rounded())" : nil
+            dateOfBirth: dateOfBirth,
+            sex: sex,
+            height: "\(height)cm",
+            weight: "\(weight)kg",
+            bmi: "\(bmi)"
         )
     }
     
-    func fetchHealth(type: String) async -> HealthData? {
+    func fetchHealth(_ healthKit: HealthKit, type: String) async throws -> HealthData? {
         guard let (identifier, _) = hkStringToHKQuantityTypeIdentifier(type),
-              let unit = identifier.siUnit else {
+              let unit = identifier.siUnit,
+              let sampleType = SampleType(identifier) else {
             return nil
         }
-        
-        let endDates = [
-            ("day", Calendar.current.date(byAdding: .day, value: -1, to: Date())!),
-            ("week", Calendar.current.date(byAdding: .day, value: -7, to: Date())!),
-            ("month", Calendar.current.date(byAdding: .month, value: -1, to: Date())!)
+
+        let timeRanges: [(String, HealthKitQueryTimeRange)] = [
+            ("day", .today),
+            ("week", .currentWeek),
+            ("month", .currentMonth)
         ]
-        
-        
-        return await withTaskGroup(of: (String, Double?).self, returning: HealthData?.self) { taskGroup in
-            for (description, endDate) in endDates {
-                taskGroup.addTask {
-                    (
-                        description,
-                        try? await self.fetchSample(
-                            for: identifier,
-                            unit: unit,
-                            startDate: endDate,
-                            endDate: .now
-                        )
```

Comment:
Currently, we’re aggregating time-series data into a single value (e.g., average heart rate over the last month). It might be more meaningful to use a different aggregation approach; for example, retrieving daily averages over the last month instead of just one number. This would also align better with future EmbedHealth integration, even if that’s a larger effort. Let me know what you think @PSchmiedmayer @LeonNissen @lukaskollmer

Reply:
agree that a higher granularity would be desirable but wasn't the issue here that having too many samples could exceed the context window? (i might be confusing this with something else).

iirc another idea of mine at the time was that we could use different granularities depending on how far back the data was collected (eg weekly heart rate for > a year ago, daily average for > last week, and hourly averages for the past 7 days)

Reply:
Yes, the main issue was the growing context window; that's why we constrained it to a few samples.

Agree with @lukaskollmer here for this approach. I think 7 days for a week, weekly samples for a month, or monthly samples for a year would make sense and don't dramatically exceed the context window. Would be good to verify this with some on-device testing @max-rosenblattl before we merge this.

Reply:
Thanks for the insights! I think it makes a lot of sense, happy to tackle that. 

Regarding testing on device: I will be able to do so in the very near future; until then if anyone has some spare minutes, I would appreciate it!

Reply:
Quick update: finally got my new iPhone and could test it on device (results in PR description above)! Also adapted the statistics query to handle the discussed time-range based aggregation intervals. Just saw they were integrated in here too https://github.com/StanfordSpezi/SpeziHealthKit/pull/65 , quite cool looking forward to the merge 🚀